### PR TITLE
add failing tests for unstable_resolve

### PR DESCRIPTION
**Pull request by dmaskasky**

### CODE REVIEW

**Comment by dai-shi** [comment]
wait, i can just merge it, can i?

File: tests/vanilla/store.test.tsx [hunk 1]
```diff
@@ -1,7 +1,9 @@
 import { waitFor } from '@testing-library/dom'
 import { assert, describe, expect, it, vi } from 'vitest'
 import { atom, createStore } from 'jotai/vanilla'
-import type { Getter } from 'jotai/vanilla'
+import type { Atom, Getter, WritableAtom } from 'jotai/vanilla'
+import test from 'node:test'
+import exp from 'node:constants'
```
Comment (dai-shi) [inline tests/vanilla/store.test.tsx:6]: what are those?

**Comment by dai-shi** [comment]
https://github.com/pmndrs/jotai/actions/runs/9231515051/job/25401510369?pr=2574 are type errors expected?

File: tests/vanilla/store.test.tsx (hunk 1 above)
Comment (dmaskasky) [reply tests/vanilla/store.test.tsx:6]: oh, I think my IDE accidentally added them. I will remove.

**Comment by dmaskasky** [comment]
> wait, i can just merge it, can i?

yes, one sec. I'll remove the extraneous imports first.

**Comment by dmaskasky** [comment]
> wait, i can just merge it, can i?

yeah, go ahead.

**Comment by dmaskasky** [comment]
> https://github.com/pmndrs/jotai/actions/runs/9231515051/job/25401510369?pr=2574 are type errors expected?

Yes

File: tests/vanilla/store.test.tsx [hunk 2]
```diff
@@ -440,3 +440,162 @@ describe('async atom with subtle timing', () => {
...
 })
+
+describe('unstable_resolve resolves the correct value for', () => {
+  function nextTask() {
+    return new Promise((resolve) => setTimeout(resolve))
+  }
+
+  it('primitive atom', async () => {
+    const store = createStore()
+    store.unstable_resolve = (atom: Atom<unknown>): Atom<unknown> => {
```
Comment (dai-shi) [inline tests/vanilla/store.test.tsx:451]: Should it be this to pass the type test?
```suggestion
    store.unstable_resolve = <T>(atom: Atom<T>): Atom<T> => {
```

Comment (dmaskasky) [reply tests/vanilla/store.test.tsx:451]: But it could return any typed atom, right?

Comment (dai-shi) [reply tests/vanilla/store.test.tsx:451]: Generally speaking, it should return the atom with the same capability, not just types.

Comment (dmaskasky) [reply tests/vanilla/store.test.tsx:451]: Ok, lets strongly type it then.

File: tests/vanilla/store.test.tsx [hunk 3]
```diff
@@ -440,3 +440,162 @@ describe('async atom with subtle timing', () => {
...
+    expect(store.get(pseudo)).toBe('a:a-mounted')
+    store.set(pseudo, (v) => v + ':a-updated')
+    expect(store.get(pseudo)).toBe('a:a-mounted:a-updated')
+    await nextTask()
+    expect(store.get(pseudo)).toBe('a:a-mounted:a-updated')
+  })
+
+  it('derived atom', async () => {
+    const store = createStore()
+    store.unstable_resolve = (atom: Atom<unknown>): Atom<unknown> => {
```
Comment (dai-shi) [inline tests/vanilla/store.test.tsx:479]:
```suggestion
    store.unstable_resolve = <T>(atom: Atom<T>): Atom<T> => {
```

File: tests/vanilla/store.test.tsx [hunk 4]
```diff
@@ -440,3 +440,162 @@ describe('async atom with subtle timing', () => {
...
+    delete store.unstable_resolve
+    await nextTask()
+    expect(store.get(pseudo)).toEqual('pseudo')
+    await nextTask()
+    expect(store.get(pseudo)).toEqual('pseudo:pseudo-mounted')
+  })
+
+  it('writable atom', async () => {
+    const store = createStore()
+    store.unstable_resolve = (atom: Atom<unknown>): Atom<unknown> => {
```
Comment (dai-shi) [inline tests/vanilla/store.test.tsx:519]:
```suggestion
    store.unstable_resolve = <T>(atom: Atom<T>): Atom<T> => {
```

File: tests/vanilla/store.test.tsx [hunk 5]
```diff
@@ -440,3 +440,162 @@ describe('async atom with subtle timing', () => {
...
+    expect(typeof value).toBe('function')
+    expect(store.get(pseudo)).toBe('a:a-mounted:a-updated')
+    unsub()
+    await nextTask()
+    expect(store.get(pseudo)).toBe('a:a-mounted:a-updated:a-unmounted')
+  })
+
+  it('this in read and write', async () => {
+    const store = createStore()
+    store.unstable_resolve = (atom: Atom<unknown>): Atom<unknown> => {
```
Comment (dai-shi) [inline tests/vanilla/store.test.tsx:551]:
```suggestion
    store.unstable_resolve = <T>(atom: Atom<T>): Atom<T> => {
```

File: tests/vanilla/store.test.tsx [hunk 6]
```diff
@@ -440,3 +440,162 @@ describe('async atom with subtle timing', () => {
...
+    const pseudo = atom('pseudo') as typeof this_read
+    pseudo.debugLabel = 'pseudo'
+    let i = 0
+    const this_read = atom(function read(this: any, get) {
+      return i++ % 2 ? get(this) : 'this_read'
+    })
+    this_read.debugLabel = 'this_read'
+    expect(store.get(pseudo)).toBe('this_read')
+
+    store.unstable_resolve = (atom: Atom<unknown>): Atom<unknown> => {
```
Comment (dai-shi) [inline tests/vanilla/store.test.tsx:567]:
```suggestion
    store.unstable_resolve = <T>(atom: Atom<T>): Atom<T> => {
```

File: tests/vanilla/store.test.tsx (hunk 3 above)
Comment (dai-shi) [inline tests/vanilla/store.test.tsx:479]:
```suggestion
    store.unstable_resolve = <T>(atom: Atom<T>): Atom<T> => {
```

File: tests/vanilla/store.test.tsx [hunk 7]
```diff
@@ -440,3 +440,162 @@ describe('async atom with subtle timing', () => {
...
 })
+
+describe('unstable_resolve resolves the correct value for', () => {
+  function nextTask() {
+    return new Promise((resolve) => setTimeout(resolve))
+  }
+
+  it('primitive atom', async () => {
+    const store = createStore()
+    store.unstable_resolve = <T>(atom: Atom<T>): Atom<T> => {
```
Comment (dai-shi) [inline tests/vanilla/store.test.tsx:451]:
```suggestion
    store.unstable_resolve = <T,>(atom: Atom<T>): Atom<T> => {
```

File: tests/vanilla/store.test.tsx [hunk 8]
```diff
@@ -440,3 +440,162 @@ describe('async atom with subtle timing', () => {
...
+    expect(store.get(pseudo)).toBe('a:a-mounted')
+    store.set(pseudo, (v) => v + ':a-updated')
+    expect(store.get(pseudo)).toBe('a:a-mounted:a-updated')
+    await nextTask()
+    expect(store.get(pseudo)).toBe('a:a-mounted:a-updated')
+  })
+
+  it('derived atom', async () => {
+    const store = createStore()
+    store.unstable_resolve = <T>(atom: Atom<T>): Atom<T> => {
```
Comment (dai-shi) [inline tests/vanilla/store.test.tsx:479]:
```suggestion
    store.unstable_resolve = <T,>(atom: Atom<T>): Atom<T> => {
```

File: tests/vanilla/store.test.tsx [hunk 9]
```diff
@@ -440,3 +440,162 @@ describe('async atom with subtle timing', () => {
...
+    delete store.unstable_resolve
+    await nextTask()
+    expect(store.get(pseudo)).toEqual('pseudo')
+    await nextTask()
+    expect(store.get(pseudo)).toEqual('pseudo:pseudo-mounted')
+  })
+
+  it('writable atom', async () => {
+    const store = createStore()
+    store.unstable_resolve = <T>(atom: Atom<T>): Atom<T> => {
```
Comment (dai-shi) [inline tests/vanilla/store.test.tsx:519]:
```suggestion
    store.unstable_resolve = <T,>(atom: Atom<T>): Atom<T> => {
```

File: tests/vanilla/store.test.tsx [hunk 10]
```diff
@@ -440,3 +440,162 @@ describe('async atom with subtle timing', () => {
...
+    expect(typeof value).toBe('function')
+    expect(store.get(pseudo)).toBe('a:a-mounted:a-updated')
+    unsub()
+    await nextTask()
+    expect(store.get(pseudo)).toBe('a:a-mounted:a-updated:a-unmounted')
+  })
+
+  it('this in read and write', async () => {
+    const store = createStore()
+    store.unstable_resolve = <T>(atom: Atom<T>): Atom<T> => {
```
Comment (dai-shi) [inline tests/vanilla/store.test.tsx:551]:
```suggestion
    store.unstable_resolve = <T,>(atom: Atom<T>): Atom<T> => {
```

File: tests/vanilla/store.test.tsx [hunk 11]
```diff
@@ -440,3 +440,162 @@ describe('async atom with subtle timing', () => {
...
+    const pseudo = atom('pseudo') as typeof this_read
+    pseudo.debugLabel = 'pseudo'
+    let i = 0
+    const this_read = atom(function read(this: any, get) {
+      return i++ % 2 ? get(this) : 'this_read'
+    })
+    this_read.debugLabel = 'this_read'
+    expect(store.get(pseudo)).toBe('this_read')
+
+    store.unstable_resolve = <T>(atom: Atom<T>): Atom<T> => {
```
Comment (dai-shi) [inline tests/vanilla/store.test.tsx:567]:
```suggestion
    store.unstable_resolve = <T,>(atom: Atom<T>): Atom<T> => {
```

File: tests/vanilla/store.test.tsx [hunk 12]
```diff
@@ -440,3 +440,162 @@ describe('async atom with subtle timing', () => {
...
 })
+
+describe('unstable_resolve resolves the correct value for', () => {
+  function nextTask() {
+    return new Promise((resolve) => setTimeout(resolve))
+  }
+
+  it('primitive atom', async () => {
+    const store = createStore()
+    store.unstable_resolve = <T,>(atom: Atom<T>): Atom<T> => {
```
Comment (dai-shi) [inline tests/vanilla/store.test.tsx:451]:
```suggestion
    store.unstable_resolve = (atom: Atom<string>): Atom<string> => {
```

**Comment by dai-shi** [comment]
Now, I'm not quite sure what would be the correct typing. We can revisit in the future. Merging for now.

**Pull request merged**